File: record_android/android/src/main/java/com/llfbandit/record/record/stream/RecorderStateStreamHandler.java
package com.llfbandit.record.record.stream;

import android.app.Activity;

import androidx.annotation.Nullable;

import io.flutter.plugin.common.EventChannel;

public class RecorderStateStreamHandler implements EventChannel.StreamHandler {
  // Event producer
  private EventChannel.EventSink eventSink;

  @Nullable
  private Activity activity;

  @Override
  public void onListen(Object o, EventChannel.EventSink eventSink) {
    this.eventSink = eventSink;
  }

  @Override
  public void onCancel(Object o) {
    eventSink = null;
  }

  public void sendStateEvent(int state) {
    if (eventSink != null && activity != null) {
      activity.runOnUiThread(() -> eventSink.success(state));
    }
  }

  public void sendStateErrorEvent(Exception ex) {
    if (eventSink != null && activity != null) {
      activity.runOnUiThread(() -> eventSink.error("-1", ex.getMessage(), ex));
    }
  }

  public void setActivity(@Nullable Activity activity) {
    this.activity = activity;
  }
}

File: record_android/android/src/main/kotlin/com/llfbandit/record/record/stream/RecorderStateStreamHandler.kt
package com.llfbandit.record.record.stream

import com.llfbandit.record.record.model.RecordState
import com.llfbandit.record.record.util.MainThread
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink

class RecorderStateStreamHandler : EventChannel.StreamHandler {
  // Event producer
  private var eventSink: EventSink? = null
  private var state: RecordState = RecordState.STOP

  override fun onListen(arguments: Any?, events: EventSink?) {
    this.eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
  }

  fun sendStateEvent(state: RecordState) {
    if (this.state != state) {
      this.state = state

      MainThread.post {
        eventSink?.success(state.id)
      }
    }
  }

  fun sendStateErrorEvent(ex: Exception) {
    MainThread.post {
      // `details` must be codec-encodable; a Throwable is not.
      eventSink?.error("-1", ex.message ?: ex.toString(), ex.cause?.toString())
    }
  }
}
